lib/datadog/statsd/connection_cfg.rb

```diff
@@ -10,28 +10,31 @@ class ConnectionCfg
       DEFAULT_PORT = 8125
 
       def initialize(host: nil, port: nil, socket_path: nil)
+        initialize_with_constructor_args(host, port, socket_path) || initialize_with_env_vars || initialize_with_defaults
+      end
 
-        # Try with constructor args
-        if setup_with(host, port, socket_path)
-          return
-        end
+      def initialize_with_constructor_args(host, port, socket_path)
+        try_initialize_with(host, port, socket_path,
+          "Both host/port and socket_path constructor arguments are set.  Set only one or the other.",
+          )
+      end
 
-        # Try with env vars
-        if setup_with(
-            ENV['DD_AGENT_HOST'],
-            ENV['DD_DOGSTATSD_PORT'].nil? ? nil : ENV['DD_DOGSTATSD_PORT'].to_i,
-            ENV['DD_DOGSTATSD_SOCKET'])
-          return
-        end
+      def initialize_with_env_vars()
+        try_initialize_with(
+          ENV['DD_AGENT_HOST'],
+          ENV['DD_DOGSTATSD_PORT'].nil? ? nil : ENV['DD_DOGSTATSD_PORT'].to_i,
+          ENV['DD_DOGSTATSD_SOCKET'],
+          "Both $DD_AGENT_HOST/$DD_DOGSTATSD_PORT and $DD_DOGSTATSD_SOCKET are set.  Set only one or the other.",
+          )
+      end
 
-        # Fall back to defaults
-        setup_with(DEFAULT_HOST, DEFAULT_PORT, nil)
+      def initialize_with_defaults()
+        try_initialize_with(DEFAULT_HOST, DEFAULT_PORT, nil, "")
       end
 
-      # set up the configuration with the given values; this is a helper for #initialize
-      def setup_with(host, port, socket_path)
+      def try_initialize_with(host, port, socket_path, not_both_error_message)
```

Comment:
Minor: When methods start having quite a few arguments, and there's even things such as `nil`, I usually suggest switching over to keyword arguments:

```ruby
def initialize_with_constructor_args(host, port, socket_path)
  try_initialize_with(
    host: host,
    port: port,
    socket_path: socket_path,
    not_both_error_message: '....',
  )
end

def initialize_with_env_vars
  try_initialize_with(
    host: ENV['DD_AGENT_HOST'],
    port: ENV['DD_DOGSTATSD_PORT'] && ENV['DD_DOGSTATSD_PORT'].to_i, # <-- smaller check
    socket_path: ENV['DD_DOGSTATSD_SOCKET'],
    not_both_error_message: '...')
  )
end

def initialize_with_defaults
  try_initialize_with # uses defaults for everything
end

def try_initialize_with(host: nil, port: nil, socket_path: nil, not_both_error_message: 'Unexpected, please report a bug')
```

Just out of curiosity, it's kind of annoying that we still need to support such old Rubies (2.0), since:

* From 2.1+ we could have used required keyword arguments, e.g. `def try_initialize_with(host:, port:, socket_path: nil, not_both_error_message: 'Unexpected, please report a bug')`
* From 2.3+ we could have used safe navigation e.g. `port: ENV['DD_DOGSTATSD_PORT']&.to_i`
* In 3.1+ will finally be able to write `try_initialize_with(host:, port:, socket_path:, ...)` instead of needing to repeat the variable name when both the key and the value are the same, e.g. `try_initialize_with(host: host, port: port, socket_path: socket_path, ...`)